app/products/calls/actions/calls.ts

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import InCallManager from 'react-native-incall-manager';
+
+import {forceLogoutIfNecessary} from '@actions/remote/session';
+import {fetchUsersByIds} from '@actions/remote/user';
+import {hasMicrophonePermission} from '@app/products/calls/actions/permissions';
+import {
+    getCallsConfig,
+    myselfJoinedCall,
+    myselfLeftCall,
+    setCalls,
+    setChannelEnabled,
+    setConfig,
+    setPluginEnabled,
+    setScreenShareURL,
+    setSpeakerPhone,
+} from '@app/products/calls/state';
+import {
+    Call,
+    CallParticipant,
+    CallsConnection, DefaultCallsConfig,
+    ServerChannelState,
+} from '@app/products/calls/types/calls';
+import Calls from '@constants/calls';
+import NetworkManager from '@managers/network_manager';
+
+import {newConnection} from '../connection/connection';
+
+import type {Client} from '@client/rest';
+import type ClientError from '@client/rest/error';
+import type {IntlShape} from 'react-intl';
+
+// Only exported for tests, not exported from the module index.
+export let connection: CallsConnection | null = null;
+
+export const loadConfig = async (serverUrl: string, force = false) => {
+    const now = Date.now();
+    const config = getCallsConfig(serverUrl);
+
+    if (!force) {
+        const lastRetrievedAt = config.last_retrieved_at || 0;
+        if ((now - lastRetrievedAt) < Calls.RefreshConfigMillis) {
+            return {data: config};
+        }
+    }
+
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+
+    let data;
+    try {
+        data = await client.getCallsConfig();
+    } catch (error) {
+        await forceLogoutIfNecessary(serverUrl, error as ClientError);
+
+        // Reset the config to the default (off) since it looks like Calls is not enabled.
+        setConfig(serverUrl, {...DefaultCallsConfig, last_retrieved_at: now});
+
+        return {error};
+    }
+
+    const nextConfig = {...config, ...data, last_retrieved_at: now};
+    setConfig(serverUrl, nextConfig);
+    return {data: nextConfig};
+};
+
+export const loadCalls = async (serverUrl: string, userId: string) => {
+    let client: Client;
+    try {
+        client = NetworkManager.getClient(serverUrl);
+    } catch (error) {
+        return {error};
+    }
+    let resp: ServerChannelState[] = [];
+    try {
+        resp = await client.getCalls();
+    } catch (error) {
+        await forceLogoutIfNecessary(serverUrl, error as ClientError);
+        return {error};
+    }
+    const callsResults: Dictionary<Call> = {};
+    const enabledChannels: Dictionary<boolean> = {};
+
+    // Batch load userModels async because we'll need them later
+    const ids = new Set<string>();
+    resp.forEach((channel) => {
+        channel.call?.users.forEach((id) => ids.add(id));
+    });
```

Comment:
Good point! Nested loops should usually be avoided. In this case I think it's okay because we're only creating a set of ids (no work is being done in the loop).

Reply:
Yes, good point, done here: 10a1236de88a3b9a5f2a4f576609adb38c4668df

Reply:
Thanks, addressed in: 10a1236de88a3b9a5f2a4f576609adb38c4668df